engine/src/cmd/damageable.cpp

```diff
@@ -267,11 +333,155 @@ void Damageable::ApplyDamage( const Vector &pnt,
+    bool armor_damage = inflicted_damage.inflicted_damage_by_layer[0] > 0;
+
+    // It's actually easier to read this condition than the equivalent form
+    if(!(hull_damage || (system_damage_on_armor && armor_damage))) {
+        return;
+    }
+
+    float& hull = layers[0].facets[0].health.health;
+    float& max_hull = layers[0].facets[0].health.factory_max_health;
+
+    bool damage_system;
+    if(player) {
+        damage_system = DestroyPlayerSystem(hull, max_hull, 1);
+    } else {
+        damage_system = DestroySystem(hull, max_hull, 1);
+    }
+
+    if(!damage_system) {
+        return;
+    }
 
+    unit->DamageRandSys( system_failure * rand01() +
+                   (1-system_failure) * ( 1-(hull > 0 ? hull_damage/hull : 1.0f) ),
+                   attack_vector, 1.0f, 1.0f );
+}
+
+void Damageable::DamageCargo(InflictedDamage inflicted_damage) {
+    // TODO: lib_damage
+    // The following code needs to be renabled and placed somewhere
+    // Non-lethal/Disabling Weapon code here
+    // TODO: enable
+    /*static float disabling_constant =
+        XMLSupport::parse_float( vs_config->getVariable( "physics", "disabling_weapon_constant", "1" ) );
+    if (hull > 0)
+      pImage->LifeSupportFunctionality += disabling_constant*damage/hull;
+    if (pImage->LifeSupportFunctionality < 0) {
+        pImage->LifeSupportFunctionalityMax += pImage->LifeSupportFunctionality;
+        pImage->LifeSupportFunctionality     = 0;
+        if (pImage->LifeSupportFunctionalityMax < 0)
+          pImage->LifeSupportFunctionalityMax = 0;
+      }*/
+
+
+    static bool system_damage_on_armor = GameConfig::GetVariable( "physics", "system_damage_on_armor", false );
+
+    bool hull_damage = inflicted_damage.inflicted_damage_by_layer[0] > 0;
+    bool armor_damage = inflicted_damage.inflicted_damage_by_layer[0] > 0;
+
+    // TODO: Same condition as DamageRandomSystem - move up and merge
+    if(!(hull_damage || (system_damage_on_armor && armor_damage))) {
+        return;
+    }
+
+    Unit *unit = static_cast<Unit*>(this);
+
+    // If nothing to damage, exit
+    if (unit->numCargo() == 0) {
+        return;
+    }
+
+    float& hull = layers[0].facets[0].health.health;
+    float& max_hull = layers[0].facets[0].health.factory_max_health;
+
+    // Is the hit unit, lucky or not
+    if ( DestroySystem( hull, max_hull, unit->numCargo() ) ) {
+        return;
+    }
+
+    static std::string restricted_items = vs_config->getVariable( "physics", "indestructable_cargo_items", "" );
+    int cargo_to_damage_index = rand() % unit->numCargo();
+    Cargo& cargo = unit->GetCargo( cargo_to_damage_index );
+    const std::string& cargo_category = cargo.GetCategory();
+
+    bool is_upgrade = cargo_category.find( "upgrades/" ) == 0;
+    bool already_damaged = cargo_category.find( "upgrades/Damaged/" ) == 0;
+    bool is_multiple = cargo_category.find( "mult_" ) == 0;
+    bool is_restricted = restricted_items.find(cargo.GetContent()) == string::npos;
```

Comment:
Hmm. It almost seems like the logic for each of these lines is reversed. But I could be wrong.

Reply:
I was sure you were right and then I found the original code and we're both wrong :). The find function returns the index of the substring. It's only is_upgrade for example, if the string starts with "upgrades/"